app/models/manageiq/providers/openstack/cloud_manager/auth_key_pair.rb

```diff
@@ -1,2 +1,23 @@
 class ManageIQ::Providers::Openstack::CloudManager::AuthKeyPair < ManageIQ::Providers::CloudManager::AuthKeyPair
+
+  def self.create_keypair(ems, name, public_key)
+    create_options = {:name => name, :public_key => public_key}
+    connection_options = {:service => 'Compute'}
+    ems.with_provider_connection(connection_options) do |service|
+      service.key_pairs.create(create_options)
+    end
+  rescue => err
+    _log.error "keypair=[#{name}], error: #{err}"
+    raise MiqException::MiqOpenstackApiRequestError, err.to_s, err.backtrace
+  end
+
+  def delete_keypair(ems)
```

Comment:
probably do a helper for the connection_options, you should always pass the correct cloud_tenant too. This tenant is used for scoping the keystone token.